modules/propertybasedtesting/include/inviwo/propertybasedtesting/testproperty.h

```diff
@@ -0,0 +1,491 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2020 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#pragma once
+
+#include <inviwo/propertybasedtesting/propertybasedtestingmoduledefine.h>
+#include <inviwo/propertybasedtesting/algorithm/propertyanalyzing.h>
+
+#include <inviwo/core/network/processornetwork.h>
+#include <inviwo/core/properties/ordinalproperty.h>
+#include <inviwo/core/properties/cameraproperty.h>
+#include <inviwo/core/properties/boolcompositeproperty.h>
+
+#include <filesystem>
+#include <fstream>
+
+namespace inviwo {
+
+class TestResult;
+
+class TestPropertyObservable;
+
+class TestPropertyObserver : public Observer {
+public:
+    friend TestPropertyObservable;
+    TestPropertyObserver() = default;
+    virtual void onTestPropertyChange() {}
+    virtual ~TestPropertyObserver() = default;
+};
+
+class TestPropertyObservable : public Observable<TestPropertyObserver> {
+protected:
+    TestPropertyObservable() = default;
+    void notifyObserversAboutChange() {
+        forEachObserver([](TestPropertyObserver* o) { o->onTestPropertyChange(); });
+    }
+    virtual ~TestPropertyObservable() = default;
+};
+
+template <typename T>
+class NetworkPath : public Serializable {
```

Comment:
We use this class for a reference to some object (Processor or Property) of type T in the ProcessorNetwork (which is
necessary since we need to access their properties or values). A normal pointer (T*) is not sufficient,
since serializing two pointers to the same object and then deserializing them yields two separate (but equal) objects, which is not what we want. Instead, we store and serialize the identifier of/path to the object (i.e. a string). Because querying the Network for the object corresponding to the identifier/path each time we want to access the referenced object seemed expensive, we also cache the pointer on the first access after deserialization.

Reply:
That seems resonable I think. What values can T have? is that always Processor or Property or can it be different things? is it always the base class or can it be derived classes to? Would be good to add this explanation as documentation of the class